### PR TITLE
Extract AMPPreactBaseElement

### PR DESCRIPTION
**Pull request by kvchari**
* creates AmpPreactBaseElement
* creates setSuperClass() to change the inheritance hierarchy of a class
* ensures all "1.0" amp-* components inherit from AmpPreactBaseElement
* moves (some) amp-specific methods into AmpPreactBaseElement
* updates amp-* component docs/scaffolding schematics to refer to AmpPreactBaseElement
* extracts CustomElement() static method out of CeBaseElement and into defineBentoElement() fn
  * ensures `defineBentoElement()` can work effectively in tests (where tests are run in an iframe)
* renames `VideoBaseElement` -> `AmpVideoBaseElement` since it's purely amp-related
* renames `BaseElement` in `amp-video` -> `BentoVideoBaseElement` to align

Partial https://github.com/ampproject/amphtml/issues/36950

### CODE REVIEW

File: src/preact/amp-base-element.js
Comment (samouri) [inline src/preact/amp-base-element.js:118]: Drive-by bikeshed (don't spend too much time here): I know @alanorozco and I disagree. But wanted to note here that I'm not a fan of the `return X` line. The code ends up looking cleaner, but also gives the appearance of purity.

Comment (kvchari) [reply src/preact/amp-base-element.js:113]: Totally get what you mean - returning X does give the illusion that a new class is created, when in reality X's prototype is being modified (which will affect all instances of X whether they realize or not)

I could change

```
class AmpFoo extends setSuperClass(BaseElement, AmpPreactBaseElement) {...}
```

to 

```
setSuperClass(BaseElement, AmpPreactBaseElement)
class AmpFoo extends BaseElement {...}
```

**Comment by samouri** [comment]
Partial for https://github.com/ampproject/amphtml/issues/36950

**Comment by kvchari** [comment]
Outstanding work to address #36950

- [ ] compress custom element wrapper + extension implementation into 1 class 
  * currently each extension is still registered with a class inheriting from `CeBaseElement` which has an `impl` property that has an instance of the extension class (e.g. AmpAccordion, BentoAccordion, etc.). The wrapper+impl pattern was done to align with amp's class structure and isn't necessary for bento components. 

Other outstanding quality of life refactors:

- [ ] rename classes representing Bento components from `BaseElement` -> `BentoFoo`
- [ ] rename files containing Bento components from `base-element.js` -> `bento-foo.js`
  * Note: currently autogenerated `bento-foo.js` files are used to publish standalone web component scripts, so there's a naming conflict here.

File: extensions/amp-embedly-card/1.0/web-component.js
Comment (samouri) [inline extensions/amp-embedly-card/1.0/web-component.js:12]: fmi: do you know why `defineElement` here defines both the card and the key, whereas `defineElementEmbedlyKey` only the key? in that case, why are they both exported

File: src/preact/amp-base-element.js
Comment (samouri) [inline src/preact/amp-base-element.js:7]: flagging for presbmit visibility

File: src/preact/amp-base-element.js
Comment (samouri) [inline src/preact/amp-base-element.js:10]: can you add this file as an override to the eslint rule? since its an amp component I don't think it needs to have this lint rule applied at all

File: src/preact/amp-base-element.js
Comment (samouri) [inline src/preact/amp-base-element.js:37]: Why are these keys strings?

File: src/preact/base-element.js
Comment (samouri) [inline src/preact/base-element.js:620]: FMI: Do bento WCs not support loading/fallback/placeholder?

**Comment by samouri** [comment]
What is `extensions/amp-accordion/1.0/test/test-bento-accordion.js` doing in this PR, and why is there an empty file called `test/unit/preact/test-amp-base-element.js`?

**Review by samouri — COMMENTED** [review]
This is looking pretty good! Left some comments/questions.

Two more:
1. What is `extensions/amp-accordion/1.0/test/test-bento-accordion.js` doing in this PR,?
2.  Why is there an empty file called `test/unit/preact/test-amp-base-element.js`? What is going to end up in that file?

File: src/preact/amp-base-element.js
Comment (kvchari) [reply src/preact/amp-base-element.js:37]: Honestly no idea - the existing `defaultProps_` is typed as a `JsonObject` which has a lint rule that it must be constructed with object literals with string props, which means you can't use spread `...{}` syntax. But I want to use that syntax to "inherit" the parent class' `defaultProps_`, so I have to disable the rule. Not sure what the preferred solution is here: disable the lint rule, change the type of the `defaultProps_` obj, etc.

File: src/preact/base-element.js
Comment (kvchari) [reply src/preact/base-element.js:620]: My understanding is that fallbacks and placeholders are an AMP concept and we're leaving the job of toggling an element on or off to consumers

@alanorozco can you confirm?

File: extensions/amp-embedly-card/1.0/web-component.js
Comment (kvchari) [reply extensions/amp-embedly-card/1.0/web-component.js:12]: > Is that true
Basically, yes

So this was necessary to allow consumers to do this:

```html
<head>
  <script async type="module" src="bento-embedly-card.js"/>
</head>
<body>
      <amp-embedly-key value={apiKey}></amp-embedly-key>
      <amp-embedly-card data-url="https://www.youtube.com/watch?v=lBTCB7yLs8Y"></amp-embedly-card>
</body>
```

the `bento-foo.js` scripts all register the "relevant" web-component(s) to the customElement registry by calling an autogenerated `defineElement` fn. When there are multiple relevant components for a particular extension, we would like them to all be registered. In those cases, we create `web-component.js` files, which override the autogenerated files built during the extension's build process, that implements a custom `defineElement` fn that registers all the necessary components.

Registering all the components together is basically what the amp components do already ([example](https://github.com/ampproject/amphtml/blob/main/extensions/amp-embedly-card/1.0/amp-embedly-card.js#L38-L41))

Although I guess it's up for debate, whether we should be exporting `defineElementEmbedlyKey` since it doesn't really make sense to use it on its own..

File: src/preact/amp-base-element.js
Comment (kvchari) [reply src/preact/amp-base-element.js:10]: What is the syntax for declare a file as "unrestricted"? Not immediately obvious from looking at [this file](https://github.com/ampproject/amphtml/blob/main/build-system/eslint-rules/restrict-this-access.js)

**Comment by kvchari** [comment]
@samouri 

> What is extensions/amp-accordion/1.0/test/test-bento-accordion.js doing in this PR,?

I was originally trying to split up the tests for amp- and bento- functionality. That file is all the modifications I had to do to isolate the "bento" functionality and test it explicitly. I gave up on splitting up tests because I hit blockers in our testing infra and wanted to talk to infra about how to resolve them. We can delete the file, but I left it in and skipped it to remind myself what we needed to do to unit test Bento components

Bento components are currently implicitly (unit) tested by the amp tests, but testing them explicitly would be necessary when we split Bento into a different repo or if the AMP component didn't inherit from the corresponding Bento component (which was 1 option I'd investigated)

> Why is there an empty file called test/unit/preact/test-amp-base-element.js? What is going to end up in that file?

I'd meant to delete that - I think that functionality should basically be tested by the amp extensions themselves?

File: extensions/amp-embedly-card/1.0/web-component.js
Comment (samouri) [reply extensions/amp-embedly-card/1.0/web-component.js:12]: > Although I guess it's up for debate, whether we should be exporting defineElementEmbedlyKey since it doesn't really make sense to use it on its own.

Please remove the export then if it doesn't make sense on its own

Comment (samouri) [reply extensions/amp-embedly-card/1.0/web-component.js:12]: Although, if this was preexisting feel free to punt to a future cleanup!

**Comment by kvchari** [comment]
> Ok! This sounds like we'd introduce separate unit tests & tackle this issue in a separate PR then right?

Yeah. Would you like me to remove this file until then?

**Comment by kvchari** [comment]
@samouri Just rebased against main to resolve conflicts with typings. Can you double check my changes to amp-base-element.js, preact/base-element.js, and bento-ce.js?

File: src/preact/base-element.js
Comment (samouri) [inline src/preact/base-element.js:564]: is this change intentional?

File: src/preact/base-element.js
Comment (samouri) [inline src/preact/base-element.js:307]: is this change intentional? do subclasses utilize `this.container_`?

File: src/preact/base-element.js
Comment (samouri) [inline src/preact/base-element.js:37]: I don't think this is right. `API_TYPE` is a generic (template) type and does not make sense on its own.

File: src/preact/bento-ce.js
Comment (samouri) [inline src/preact/bento-ce.js:90]: I don't think we should need the ts-ignore.
I think in this case you can use `devAssert(win)` to inform below code it isn't null. That or use `_win`

Comment (kvchari) [reply src/preact/bento-ce.js:90]: i wasn't ignoring a null error, I was ignoring this:

`Property 'HTMLElement' does not exist on type 'Window'.`

because HTMLElement doesn't exist on `window`'s type?.

I was able to get around this by doing this:

```
/** @type {typeof self} */
let win;
...
ExtendableHTMLElement = maybeWrapNativeSuper(win.HTMLElement);
```

Apparently `HTMLElement` does exist on `self`'s type, so I don't get the same error.

File: src/preact/base-element.js
Comment (kvchari) [reply src/preact/base-element.js:564]: Yes it's used It's used in mutatedAttributesCallback [here](https://github.com/ampproject/amphtml/pull/37309/files#diff-7d1d95d912247f7142f4a510777181aff5dfc1975f5ad4de225cd2dc6b064015R19)

File: src/preact/base-element.js
Comment (kvchari) [reply src/preact/base-element.js:307]: It's used in mutatedAttributesCallback [here](https://github.com/ampproject/amphtml/pull/37309/files#diff-7d1d95d912247f7142f4a510777181aff5dfc1975f5ad4de225cd2dc6b064015R17)

Comment (samouri) [reply src/preact/base-element.js:308]: carrry on then :)

File: src/preact/bento-ce.js
Comment (samouri) [reply src/preact/bento-ce.js]: Can you create a new file: `src/preact/global.d.ts` with the contents:
```
export {};

declare global {
  interface Window {
    HTMLElement: typeof HTMLElement;
  }
}
```

Hopefully we can figure out something better later.

File: src/preact/amp-base-element.js
Comment (samouri) [reply src/preact/amp-base-element.js:49]: Given Alan's PR, can skip `dict()` entirely

File: src/preact/bento-ce.js
Comment (kvchari) [reply src/preact/bento-ce.js]: resolved in dms

we're going to use `typeof globalThis` as the type for `win`

**Review by samouri — APPROVED** [review]
Still need to remove `test/unit/preact/test-amp-base-element.js`, otherwise LGTM

File: src/preact/amp-base-element.js
Comment (kvchari) [reply src/preact/amp-base-element.js]: disabling for the whole file since it's not configurable. Can make it configurable in the future if necessary but this might be a rare need

File: src/preact/amp-base-element.js
Comment (kvchari) [reply src/preact/amp-base-element.js]: removed

File: extensions/amp-embedly-card/1.0/web-component.js
Comment (kvchari) [reply extensions/amp-embedly-card/1.0/web-component.js:12]: adding to checklist [here](https://github.com/ampproject/amphtml/pull/37309#issuecomment-1011222406)

File: src/preact/amp-base-element.js
Comment (dethstrobe) [inline src/preact/amp-base-element.js:128]:
```suggestion
export function setSuperClass(X, Y = AmpPreactBaseElement) {
```
if we default Y to `AmpPreactBaseElement` then we don't need to always import `AmpPreactBaseElement` into our amp classes, and we can always override it for the few cases where we need to change the prototype inheritance.

**Review by dethstrobe — APPROVED** [review]

File: src/preact/amp-base-element.js
Comment (samouri) [reply src/preact/amp-base-element.js:128]: I had a similar question that I was planning on asking later.
How many times are we calling setSuperClass with the same arguments? Ideally this is something that only needs to happen once.

Comment (kvchari) [reply src/preact/amp-base-element.js:128]: I agree that we should never call `setSuperClass` with the same arguments. Below are all the usages of `setSuperClass()`. I believe that `setSuperClass()` is never called twice with the exact same arguments. However, we call it 22 times with `AmpPreactBaseElement` and 6 times with `AmpVideoBaseElement` as the 2nd argument. (In all of these usages, `BaseElement` refers to different - bento - elements with the same generic name)

Re @dethstrobe's suggesion: I agree that it would look cleaner to maybe use a default for the 2nd parameter, but I also agree that this would further obfuscate the code make the more magical. Honestly my preference would be to keep it how it is (or make it not return as suggested [here](https://github.com/ampproject/amphtml/pull/37309#discussion_r779919449)) so we avoid magic. 


```js
class Amp__component_name_pascalcase__ extends setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpAccordion extends                     setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpBaseCarousel extends                  setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpDateCountdown extends                 setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpDateDisplay extends                   setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpEmbedlyCard extends                   setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpFacebookBase extends                  setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpFitText extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpIframe extends                        setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpInlineGallery extends                 setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpInstagram extends                     setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpLightbox extends                      setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpLightboxGallery extends               setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpMathml extends                        setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpRender extends                        setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSelector extends                      setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSidebar extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSocialShare extends                   setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSoundcloud extends                    setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpStreamGallery extends                 setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpTimeago extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpTwitter extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpWordPressEmbed extends                setSuperClass(BaseElement, AmpPreactBaseElement) {...}

class AmpBrightcove extends                    setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpDailymotion extends                   setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpJwplayer extends                      setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpVideoIframe extends                   setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpVimeo extends                         setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpYoutube extends                       setSuperClass(BaseElement, AmpVideoBaseElement) {...}

class AmpEmbedlyKey extends                    setSuperClass(EmbedlyKeyBaseElement, AmpPreactBaseElement) {...}
class AmpFacebookComments extends              setSuperClass(CommentsBaseElement, AmpFacebookBase) {}
class AmpFacebookLike extends                  setSuperClass(LikeBaseElement, AmpFacebookBase) {}
class AmpFacebookPage extends                  setSuperClass(PageBaseElement, AmpFacebookBase) {}
class AmpInlineGalleryPagination extends       setSuperClass(PaginationBaseElement, AmpPreactBaseElement) {...}
class AmpInlineGalleryThumbnails extends       setSuperClass(ThumbnailsBaseElement, AmpPreactBaseElement) {...}
class AmpVideoBaseElement extends              setSuperClass(BentoVideoBaseElement, AmpPreactBaseElement) {...}
```

Comment (dmanek) [inline src/preact/amp-base-element.js:128]: Can you please rename `X` and `Y` to something meaningful like `child`/`newParent` or `derived`/`newBase`?

**Review by dmanek — APPROVED** [review]

File: src/preact/amp-base-element.js
Comment (samouri) [reply src/preact/amp-base-element.js:128]: klass, superClass

**Review by estherkim — APPROVED** [review]
build system changes lgtm

**Comment by samouri** [comment]
@kvchari: anything holding up this PR?

File: src/preact/amp-base-element.js
Comment (kvchari) [reply src/preact/amp-base-element.js:128]: forgot to push this before i merged so here's a diff PR: https://github.com/ampproject/amphtml/pull/37432

i went with `clazz` and `superClazz`

**Comment by samouri** [comment]
@kvchari: are you tracking the [followup tasks](https://github.com/ampproject/amphtml/pull/37309#issuecomment-1011222406) anywhere?
I'd be happy to take a stab at one of them

**Comment by kvchari** [comment]
@samouri made an issue for them here: https://github.com/ampproject/amphtml/issues/37538

**Pull request merged**
